### PR TITLE
S3-style links bug fix

### LINKED ISSUE #56978: s3-style URLs don't work
https://pastila.nl/?00a139a3/0b483a6142a9ba07dfcec1c32d6b0fed#3YTUZegWa3J6LQrl0/JOlw==

### PR DESCRIPTION
**Pull request by yariks5s**
### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in an official stable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fix a bug when user have an s3-style link with directories.
Use case: [https://pastila.nl/?00a139a3/0b483a6142a9ba07dfcec1c32d6b0fed#3YTUZegWa3J6LQrl0/JOlw==](https://pastila.nl/?00a139a3/0b483a6142a9ba07dfcec1c32d6b0fed#3YTUZegWa3J6LQrl0/JOlw==). Closes #56978

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

### CODE REVIEW

**Comment by robot-ch-test-poll1** [comment]
*This is an automated comment for commit be3b1f818802e285788a0c6f27908a75d22cf71e with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/57075/be3b1f818802e285788a0c6f27908a75d22cf71e/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td><a href="https://github.com/ClickHouse/ClickHouse/actions/runs/6948319442">✅ success</a></td></tr>
<tbody>
</table>
</details>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/57075/be3b1f818802e285788a0c6f27908a75d22cf71e/ci_running.html">⏳ pending</a></td></tr>
<tbody>
</table>

File: src/IO/S3/URI.cpp [hunk 1]
```diff
@@ -26,7 +26,7 @@ struct URIConverter
     static void modifyURI(Poco::URI & uri, std::unordered_map<std::string, std::string> mapper)
     {
         Macros macros({{"bucket", uri.getHost()}});
-        uri = macros.expand(mapper[uri.getScheme()]).empty()? uri : Poco::URI(macros.expand(mapper[uri.getScheme()]) + "/" + uri.getPathAndQuery());
+        uri = macros.expand(mapper[uri.getScheme()]).empty()? uri : Poco::URI(macros.expand(mapper[uri.getScheme()]) + (uri.getPathAndQuery()[0] == '/' ? "" : "/") + uri.getPathAndQuery());
```
Comment (nickitat) [inline src/IO/S3/URI.cpp:29]: `URI` has the ctor of two parameters or maybe with the help of any of its other methods we could avoid manually dealing with slashes?

Comment (yariks5s) [reply src/IO/S3/URI.cpp]: I removed unused check and decided not do add any slashes at all (they are all in `getPathAndQuery()`), it is even simpler now

File: src/IO/S3/URI.cpp [hunk 2]
```diff
@@ -26,7 +26,7 @@ struct URIConverter
     static void modifyURI(Poco::URI & uri, std::unordered_map<std::string, std::string> mapper)
     {
         Macros macros({{"bucket", uri.getHost()}});
-        uri = macros.expand(mapper[uri.getScheme()]).empty()? uri : Poco::URI(macros.expand(mapper[uri.getScheme()]) + "/" + uri.getPathAndQuery());
+        uri = macros.expand(mapper[uri.getScheme()]).empty()? uri : Poco::URI(macros.expand(mapper[uri.getScheme()]) + uri.getPathAndQuery());
```
Comment (alexey-milovidov) [inline src/IO/S3/URI.cpp:29]: Missing a whitespace before `?`

Comment (yariks5s) [reply src/IO/S3/URI.cpp]: done

**Review by nickitat — APPROVED** [review]

**Pull request merged**